show style settings in a dialog

tophf opened the pull request:
![image](https://user-images.githubusercontent.com/1310400/147405849-a97abb2a-1074-4e0f-a7cb-b52b03711094.png)

Most changes are trivial moves/renames.

* Tab bar is reverted
* `Settings` button is added
* `event-emitter` is embedded in a simplified form
* `styleChange` event is `styleUpdated` now to match `method`. Its combined handler for both editors is in `base`.
* `onStyleUpdated` is `updateClass` now to differentiate it from the events
* When textarea is not empty the dialog is wider
* `can-close-on-esc` handling was generalized to allow setting this class on a parent tag

narcolepticinsomniac commented:
LGTM in a dialog. I wasn't a fan of sticking em in a tab either - seemed unnecessary. 

The info icon has an irrelevant click event. I didn't test any actual functionality beyond that, but whenever anyone asks my opinion, I'm gonna reiterate that includes and excludes need to be as customizable as they are in typical 'applies-to' rules. Once we allow for user inputted rules, domain and path aren't really going to cut it.

tophf commented:
> irrelevant click event

Had my doubts too but all other such icons are clickable...

> includes and excludes need to be as customizable as they are in typical 'applies-to' rules

Guess I agree, although it's a separate thing to implement.

narcolepticinsomniac commented:
> all other such icons are clickable

IDC whether they have click actions or not, but the dialog message I'm getting is not relative to this info icon. For short messages, hover titles are enough IMO.

tophf commented:
What about moving it to the button itself? The discoverability is lower but eventually people should get it.

![image](https://user-images.githubusercontent.com/1310400/147454934-41d3f9b1-e9aa-4d76-b7dc-3bacb3cf4a6a.png)

tophf commented:
Another thing. We should disambiguate Options from Settings by either renaming the latter to `Style settings`, which will move the button to a separate line in more languages (NBD though), or renaming the former to `Editor settings` or just `Editor` - which one?

eight04 commented on edit/base.js:
```diff
@@ -48,6 +63,19 @@ const editor = Object.assign(EventEmitter(), {
+    delete newStyle.sections;
+    delete newStyle.name;
+    delete newStyle.enabled;
+    Object.assign(editor.style, newStyle);
```
IIRC, the editor doesn't always use `editor.style` so I implemented them separately and added an event listener.

tophf replied:
It's assigned in base.js::loadStyle which runs before edit.js can emit `styleUpdated`.

eight04 commented on edit/base.js:
```diff
@@ -48,6 +63,19 @@ const editor = Object.assign(EventEmitter(), {
   },
+};
+
+editor.on('styleUpdated', (newStyle, reason) => {
```
I used `styleChange` intentionally to distinguish them from `styleUpdated`, especially as a reminder for us to avoid infinite loop. The event may be generate by `sections-editor.js`, `source-editor.js`, or `edit.js`.

tophf replied:
Using synonyms is wrong because its meaning is the same as the original event. Like we forgot about the first one and added another one. I don't understand why we need a reminder, but if we do, it should be an *explicit* one e.g. a comment on the event or an *entirely* different name.

tophf replied:
Hey, I think I kinda get it now so maybe we can differentiate by using a different way to invoke it e.g. a Proxy on `emit`...

eight04 replied:
Another method is to avoid emitting `styleUpdated` from `sections-editor.js` and `source-editor.js`. Use a new event like `styleSaved`. Therefore `styleUpdated` is always emitted from `styleMan`.

We may not need the `styleSaved` event at all. We use that event to enable the setting page after a new style is created and saved. Because the settings page is opened via a button in this PR, we can just hide the button via CSS
```css
.is-new-style #style-cfg-btn {
  display: none;
}
```

tophf replied:
> we can just hide the button

Great idea! I've implemented it.

eight04 commented on edit/base.js:
```diff
@@ -48,6 +63,19 @@ const editor = Object.assign(EventEmitter(), {
+    delete newStyle.enabled;
+    Object.assign(editor.style, newStyle);
+    editor.updateLivePreview();
+  } else if (reason !== 'new') {
```
Suppose L74 works fine, we should probably use a cleaner flow. The current event flow looks like this:
```
editor.save  -> styleChange(new) -> settings
             -> livePreview
styleUpdated -> styleChange(any)     -> settings
                styleChange(config)  -> livePreview
                styleChange(non-new) -> replaceStyle
```
It might be better to change it to
```
editor.save      -> livePreview
editor.save(new) -> settings
styleUpdated(any) -> settings
                  -> replaceStyle
styleUpdated(config) -> livePreview
```
i.e. we don't need this listener.

tophf replied:
I didn't try to change the flow, just couldn't resist removing the extra returns in such a short function. I look at the flow in your comment and I can't understand the difference just by looking so I'd say this is something that you can do in a separate PR rather than explain it to me. I'll get it when I look at the code, hopefully :-)

eight04 reviewed (commented):
Haven't test it, just look through the code.

tophf commented:
I also added the save button + autosave checkbox from our var config dialog because 1) calling API for every typed character is bad, 2) calling API for `change` event that occurs only when exiting the input is bad as the changes may be lost if the tab navigates or closes, 3) using two different types of autosave workflow is wrong, we should use just one, especially if we combine the settings with the config dialog someday.

Also simplified css/html in settings.
